### PR TITLE
Extract web authentication logic

### PR DESCRIPTION
**Pull request by koke**
This is the first step towards a simpler, more modern, web view controller. In this PR, I've extracted all the authentication logic to a separate `WebViewAuthenticator` class that has unit tests.

Some decisions that I made that might need explanation:

We still can't authenticate Jetpack sites, but we're treating them as WordPress.com for authentication purposes, because that at least enables SSO to work. It's not the ideal experience, but it's a one tap login, and then the site will be authenticated in subsequent views until the cookies expire.

In some places that are not site related (stats, reader, notifications) we're pre-authenticating with the account. The rationale is that any of those links could be a WordPress.com site even without a `.wordpress.com` domain. This means that we're also doing login when someone visits a link to an external site (e.g. github.com). WordPress.com login won't allow an open redirect, so I had to come up with a hack: it'll redirect to `https://wordpress.com/?wpios_redirect=<the_actual_redirect_to>`, and we intercept that in the web view delegate. Not 100% happy with the solution, but it was either that, or adding a lot of logic to figure out if something was a WordPress.com URL (not 100% accurate).

Improves: #7761 as the theme detail is logged in, but activating doesn't take you back to the app.
Improves: #7778 #7759 as you still get the login form, but if there's a SSO you are already logged in to WordPress.com
Related: #7781

To test:

Verify that all instances of WPWebViewController work as expected. Some examples:

- Click a link on the reader, comment, stats
- View Site
- Set up a new sharing connection for a site

To help with testing/debugging, if you want to clear cookies, you can pause the app, and use this in the debugger console:

```
(lldb) po [[[NSHTTPCookieStorage sharedHTTPCookieStorage] cookies] count]
6

(lldb) po [[NSHTTPCookieStorage sharedHTTPCookieStorage] removeCookiesSinceDate:[NSDate dateWithTimeIntervalSinceReferenceDate:0]]
(lldb) po [[[NSHTTPCookieStorage sharedHTTPCookieStorage] cookies] count]
<nil>
```

Needs review: @elibud

### CODE REVIEW

File: WordPress/Classes/Utility/WebViewAuthenticator.swift [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    func unauthenticatedRequest(url: URL) -> URLRequest {
+        return URLRequest(url: url)
+    }
+
+    func body(url: URL) -> Data? {
+        guard let encodedUsername = username.urlFormEncoded,
+            let encodedUrl = redirectUrl(url: url.absoluteString)?.urlFormEncoded else {
+            return nil
+        }
+        let encodedPassword = password?.urlFormEncoded
```
Comment (elibud) [inline WordPress/Classes/Utility/WebViewAuthenticator.swift]: This let could be moved to line 139, right?

File: WordPress/Classes/Utility/WebViewAuthenticator.swift [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+        var parameters = "log=\(encodedUsername)"
+        if let encodedPassword = encodedPassword {
+            parameters += "&pwd=\(encodedPassword)"
+        }
+        parameters += "&redirect_to=\(encodedUrl)"
+
```
Comment (elibud) [inline WordPress/Classes/Utility/WebViewAuthenticator.swift]: I don't really like composing query strings manually, what do you think about using `URLComponents` ?

```
    var queryItems = [URLQueryItem]()
    queryItems.append(URLQueryItem(name: "log", value: encodedUsername))
    if let encodedPassword = password?.urlFormEncoded {
         queryItems.append(URLQueryItem(name: "pwd", value: encodedPassword))
    }
    queryItems.append(URLQueryItem(name: "redirect_to", value: encodedUrl))
    var components = URLComponents()
    components.queryItems = queryItems

    return components.query?.data(using: .utf8)
```

File: WordPress/Classes/Utility/WebViewAuthenticator.swift [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    func redirectUrl(url: String) -> String? {
+        guard case .dotCom = credentials else {
+            return url
+        }
+        guard let encodedUrl = url.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
+            return nil
+        }
+        return "https://wordpress.com/?\(WebViewAuthenticator.redirectParameter)=\(encodedUrl)"
```
Comment (elibud) [inline WordPress/Classes/Utility/WebViewAuthenticator.swift]: Same here, can we consider using `URLComponents` or do you think it's overkill?

File: WordPress/Classes/Utility/WebViewAuthenticator.swift (hunk 2 above)
Comment (koke) [reply WordPress/Classes/Utility/WebViewAuthenticator.swift]: > what do you think about using URLComponents ?

I think that I can't believe I missed that when that's what I used for decoding the wpios_redirect parameter 😮 Thanks for the suggestion.

**Comment by koke** [comment]
> the first time I went to the theme customizer on a Jetpack site with SSO enabled, after clicking on Log in with WordPress.com I was redirected to wp-admin and not the correct customize url

I tested this again and it's working for me. Do you have another Jetpack site with SSO to test it? I'm not sure what's going on. Can you log the URLs the web view is trying to load?

I pushed fixes for the suggestions. I'm now using `URLComponents`, the encoding is slightly different but it seems to be working well. Ready for another look

**Comment by elibud** [comment]
The flow that's not working for me is actually Try & Customize (not the currently active theme). For some reason, Customize for the installed theme works fine, but Try & Customize doesn't, but they both use the same URL as far as I can see, it might be something with the redirects? 

Everything else works perfectly.

**Comment by koke** [comment]
Try&Customize also works for me in a fresh install 😕

**Comment by elibud** [comment]
OK maybe there's something wrong with my site, I've been doing awful things to it. I've sent you an invite so you can see for yourself but I'm approving this now as all the other testing is fine.

**Pull request merged**